engine/classes/Elgg/Upgrades/PrivateAndFriendAcls.php

```diff
@@ -0,0 +1,199 @@
+<?php
+
+namespace Elgg\Upgrades;
+
+class PrivateAndFriendAcls {
+	private $db;
+
+	private $accessCollections;
+
+	private $entityTable;
+
+	/**
+	 *
+	 */
+	public function __construct() {
+		$this->db = _elgg_services()->db;
+		$this->entityTable = _elgg_services()->entityTable;
+		$this->accessCollections = _elgg_services()->accessCollections;
+		$this->metastringsTable = _elgg_services()->metastringsTable;
+	}
+
+	/**
+	 *
+	 */
+	public function run(int $offset) {
+
+		$dbprefix = $this->db->getTablePrefix();
+
+		$access_private = ACCESS_PRIVATE;
+		$access_friends = ACCESS_FRIENDS;
+
+		$users = $this->entityTable->getEntities(array(
+			'type' => 'user',
+			'offset' => $offset,
+			'limit' => 50,
+		));
+
+		foreach ($users as $user) {
+			$friends_acl_id = $this->getFriendsAclId();
+			$private_acl_id = $this->getPrivateAclId();
+
+			if (!$friends_acl_id || !$private_acl_id) {
+				$error_count++;
+				// TODO Register user facing error
+				continue;
+			}
+
+			$user->friends_acl = $friends_acl_id;
+			$user->private_acl = $private_acl_id;
+
+			// Add user to their own private acl
+			$this->accessCollections->addUser($user->guid, $private_acl_id);
+
+			// Add user to their own friends collection so they still have access
+			// if ownership changes on things assigned to their friends
+			$this->accessCollections->addUser($user->guid, $friends_acl_id);
+
+			// Add missing friends to the ACL
+			$params = array(
+				'type' => 'user',
+				'relationship' => 'friend',
+				'relationship_guid' => $user->guid,
+				'limit' => false,
+				'callback' => false
+			);
+
+			$friends = new \ElggBatch('elgg_get_entities_from_relationship', $params);
+			foreach ($friends as $friend) {
+				$this->accessCollections->addUser($friend->guid, $friends_acl_id);
+			}
+
+			// update their content
+			$params = array(
+				'owner_guid' => $user->guid,
+				'wheres' => array(
+					"access_id IN ({$access_private}, {$access_friends})"
+				),
+				'limit' => false
+			);
+
+			$content = new \ElggBatch('elgg_get_entities', $params, null, 25, false);
+			foreach ($content as $entity) {
+				// Skip comments and discussion replies as their ACLs are
+				// updated to match their containers' ACLs automatically
+				if ($entity instanceof ElggComment) {
```

Comment:
Just noticed we can't be skipping things here in an elggbatch with offset false - we'll get an infinite loop once the skipped entities exceed the batch iteration.  Need to exclude them in the query.